Update documentation for arbitrary_enum_discriminant feature (#639)
rust-lang/reference: src/items/enumerations.md

```

In this example, `Cat` is a _struct-like enum variant_, whereas `Dog` is simply

> called an enum variant.

Reviewer: The re-organization looks fine, but we lose our definition of a field-less enum. This suggestion re-adds it. I don't actually know if the anchor actually works. It should just create an anchor that can be linked too without changing any styles or making it clickable.

```suggestion
called an enum variant.

An enum where no constructors contain fields are called a *<a name="field-less-enum">field-less enum</a>*.
```

## Discriminants

Each enum instance has a _discriminant_: an integer logically associated to it

> that is used to determine which variant it holds. An opaque reference to this

Reviewer: Using `mem::discriminant` should be in the accessing the discriminant section.

discriminant can be obtained with the [`mem::discriminant`] function.

Under the [default representation], the discriminant is interpreted as
an `isize` value. However, the compiler is allowed to use a smaller type (or
another means of distinguishing variants) in its actual memory layout.

> If the [primitive representation] or the [`C` representation] is used, the

Reviewer: I don't understand what this paragraph is trying to say.

> leading bytes of a variant (e.g., two bytes if `#[repr(u16)]` is used), will

Reviewer: ```suggestion
leading bytes of a variant (for example, two bytes if `#[repr(u16)]` is used), will
```

correspond exactly to the discriminant.

> ### Assigning Discriminant Values

Reviewer: For this entire section, it only makes sense for field-less enums or enums with a primitive representation (or the `C` representation?). The way it is written, it seems like explicit discriminants are only for those and implicit applies to all. Instead, it should state that having a non-opaque (feel free to pick a different descriptor) discriminant applies in those situations and in all others, it is opaque and must only be accessed via `mem::discriminant`.

#### Explicit Discriminants

In two circumstances, the discriminant of a variant may be explicitly set by
following the variant name with `=` and a [constant expression]:

<ol>
<li>

> if the enumeration is "C-like" (i.e., it has no tuple or struct variants); e.g.:

Reviewer: ```suggestion
if the enumeration is field-less. For example:
```

Author: Per https://github.com/rust-lang-nursery/reference/pull/639#issuecomment-511515290, these terms may not be truly interchangeable.

```rust

> # #![feature(arbitrary_enum_discriminant)]

Reviewer: Feature flags aren't needed; the prose will only be merged after the feature is stable.

```suggestion
```

enum Enum {
    Foo = 3,
    Bar = 2,
    Baz = 1,
}
```
</li>
<li>

> if a [primitive representation] is used; e.g.:

Reviewer: ```suggestion
if a [primitive representation] is used. For example:
```

```rust

> # #![feature(arbitrary_enum_discriminant)]

Reviewer: Feature flag not needed.

```suggestion
```

}
```

### Accessing Discriminant Values

#### Casting

> If there is no data attached to *any* of the variants of an enumeration, then

Reviewer: There's no need to describe a field-less enum here. Just use the term.

```suggestion
If the enum is a field-less enum, then
```

Author: Per https://github.com/rust-lang-nursery/reference/pull/639#issuecomment-511515290, these terms may not be completely interchangeable. If "field-less" is meant to be synonymous to "C-like", then this change would be incorrect.
